src/raftstore/coprocessor/region_snapshot.rs

```diff
@@ -38,77 +38,42 @@ impl<'a> RegionSnapshot<'a> {
         }
     }
 
-    pub fn from_raw(db: &'a DB, region: metapb::Region) -> RegionSnapshot<'a> {
+    pub fn from_raw(db: &'a DB, region: Region) -> RegionSnapshot<'a> {
         RegionSnapshot {
             snap: db.snapshot(),
             region: region,
         }
     }
 
-    fn new_iterator(&'a self, start_key: &[u8]) -> Box<Iterator<Item = Kv> + 'a> {
-        let scan_start_key = if start_key < self.region.get_start_key() {
-            keys::data_key(self.region.get_start_key())
-        } else {
-            keys::data_key(start_key)
-        };
-        let scan_end_key = enc_end_key(&self.region);
-        box self.snap
-            .new_iterator(&scan_start_key)
-            .take_while(move |&(k, _)| k < &scan_end_key)
-            .map(|(k, v)| (keys::origin_key(k), v))
-    }
-
-    // Seek the first key >= given key, if no found, return None.
-    pub fn seek(&self, key: &[u8]) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
-        let pair = self.new_iterator(key)
-            .next()
-            .map(|(k, v)| (k.to_vec(), v.to_vec()));
-        Ok(pair)
-    }
-
-    fn new_reverse_iterator(&'a self, start_key: &[u8]) -> Box<Iterator<Item = Kv> + 'a> {
-        let scan_start_key = if start_key > self.region.get_end_key() &&
-                                !self.region.get_end_key().is_empty() {
-            enc_end_key(&self.region)
-        } else {
-            keys::data_key(start_key)
-        };
-        let mut iter = self.snap.iterator(IteratorMode::From(&scan_start_key, Direction::Reverse));
-        if !iter.valid() {
-            iter = self.snap.iterator(IteratorMode::End);
-        }
-        let scan_end_key = keys::data_key(self.region.get_start_key());
-        box iter.skip_while(move |&(k, _)| k >= &scan_start_key)
-            .take_while(move |&(k, _)| k >= &scan_end_key)
-            .map(|(k, v)| (keys::origin_key(k), v))
-    }
-
-    // Seek the first key < given key, if no found, return None.
-    pub fn reverse_seek(&self, key: &[u8]) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
-        let pair = self.new_reverse_iterator(key).next().map(|(k, v)| (k.to_vec(), v.to_vec()));
-        Ok(pair)
+    pub fn get_region(&self) -> &Region {
+        &self.region
     }
 
-    pub fn get_region(&self) -> &metapb::Region {
-        &self.region
+    pub fn iter(&self) -> RegionIterator {
+        let start_key = keys::enc_start_key(&self.region);
+        RegionIterator::new(self.snap.iter(start_key.as_slice().into()),
```

Comment:
can we implement from Vec<u8> to SeekKey too? I see many places use as_slice().into()